### PR TITLE
[dynamo] Add a `DynamoFrameType` type above Python frame object

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140155
* #140154
* #140153
* #140435
* #140436
* #140152
* __->__ #140330

This patch introduces a `DynamoFrameType` to serve as a layer between
Dynamo and different versions of Python frame object. In
`DynamoFrameType`, we only register attributes Dynamo cares about (e.g.,
`f_code`, `f_locals`, etc.

This will be helpful when it comes to adding new attributes to this
`DynamoFrameType`, or dealing with Python version changes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/csrc/dynamo/eval_frame.c [hunk 1]
```diff
@@ -78,22 +82,20 @@ DECLARE_PYOBJ_ATTR(f_executable)
...
 
-DECLARE_PYOBJ_ATTR(frame_obj)
-
 #undef DECLARE_PYOBJ_ATTR
 
-static THPPyInterpreterFrame* THPPyInterpreterFrame_previous(THPPyInterpreterFrame* self, PyObject* _noargs) {
-  THPPyInterpreterFrame* res = THPPyInterpreterFrame_New(self->frame->previous);
-  return res;
-}
-
```
Comment (StrongerXi) [inline torch/csrc/dynamo/eval_frame.c:89]: I removed these because I couldn't find where they are used, maybe CI will tell us.

**Comment by StrongerXi** [comment]
Rebase

**Review by jansel — APPROVED** [review]

File: torch/csrc/dynamo/eval_frame.c [hunk 2]
```diff
@@ -39,16 +39,26 @@ static void eval_frame_callback_set(PyObject* obj) {
 #if IS_PYTHON_3_11_PLUS
 #define THP_EVAL_API_FRAME_OBJECT _PyInterpreterFrame
 
+#else
+// Things specific to PY < 3.11
+#define THP_EVAL_API_FRAME_OBJECT PyFrameObject
+static int
+THP_PyFrame_FastToLocalsWithError(THP_EVAL_API_FRAME_OBJECT *frame, int *free_vars_copied) {
```
Comment (williamwen42) [inline torch/csrc/dynamo/eval_frame.c:46]: I deleted this function recently - please remove.

**Review by williamwen42 — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/eval_frame.c (hunk 2 above)
Comment (StrongerXi) [reply torch/csrc/dynamo/eval_frame.c:46]: Nice, will update.

**Review by williamwen42 — APPROVED** [review]

**Pull request closed without merging**